lib/firmware/cosmos.c

```diff
@@ -0,0 +1,307 @@
+#include "keepkey/firmware/cosmos.h"
+
+#include "keepkey/board/confirm_sm.h"
+#include "keepkey/board/util.h"
+#include "keepkey/firmware/home_sm.h"
+#include "keepkey/firmware/storage.h"
+#include "trezor/crypto/ecdsa.h"
+#include "trezor/crypto/memzero.h"
+#include "trezor/crypto/segwit_addr.h"
+#include <stdbool.h>
+#include <time.h>
+
+static CONFIDENTIAL HDNode node;
+static SHA256_CTX ctx;
+static bool has_message;
+static bool initialized;
+static uint32_t address_n[8];
+static size_t address_n_count;
+static uint32_t msgs_remaining;
+static uint64_t sequence;
+
+bool cosmos_path_mismatched(const CoinType *_coin,
+                            const uint32_t *_address_n,
+                            const uint32_t _address_n_count)
+{
+    // m/44' : BIP44-like path
+    // m / purpose' / bip44_account_path' / account' / x / y
+    bool mismatch = false;
+    mismatch |= _address_n_count != 5;
+    mismatch |= _address_n_count > 0 && (_address_n[0] != (0x80000000 + 44));
+    mismatch |= _address_n_count > 1 && (_address_n[1] != _coin->bip44_account_path);
+    mismatch |= _address_n_count > 2 && (_address_n[2] & 0x80000000) == 0;
+    mismatch |= _address_n_count > 3 && (_address_n[3] & 0x80000000) != 0;
```

Comment:
```suggestion
    mismatch |= _address_n_count > 3 && _address_n[3]  != 0;
```

Reply:
If we leave this too flexible, there's a chance of ransom attacks like what was disclosed in https://thecharlatan.github.io/Ransom-Coldcard/